### ISSUE TITLE
Does Kontext training support multi-GPU?

### ISSUE THREAD

**Issue by mollyleeeee** [kohya-ss/musubi-tuner#519]
Here is my training script:
accelerate launch --num_processes=4 --num_cpu_threads_per_process=8 $launch_args flux_kontext_train_network.py
  --dataset_config="$dataset_config" `
  --dit=$dit `
  --vae=$vae `
  --seed=$seed  `
  --learning_rate=$lr `
  --output_name=$output_name `
  --output_dir="./output_dir" `
  --logging_dir="./logs" `
  $ext_args

I think FluxKontextNetworkTrainer support accelerator training.

<img width="834" height="316" alt="Image" src="https://github.com/user-attachments/assets/d081ec39-2d07-4755-9f5f-9d69c99203ab" />
 
But When training with 2 GPU, the training time is similar to that of a 1 GPU. Although the batch size doubles, the training time per step also doubles.

2GPU：15:56:12, **14.43s/it**, avr_loss=0.356]
1GPU：14:56:54,  **7.16s/it**, lr: 1.0e-04 loss: 0.376

**Comment by kohya-ss** [comment]
DDP Multi-GPU training is supported. The time per step should not be much different from that of a single GPU.
Could you please share your training settings?

**Comment by CaiNiaoCYY** [comment]
<img width="1486" height="356" alt="Image" src="https://github.com/user-attachments/assets/34d114e5-da6e-4b75-a3a4-976f9726423f" />
so readme is Outdated？ can you give guide for multi-GPU training?

**Comment by mollyleeeee** [comment]
> DDP Multi-GPU training is supported. The time per step should not be much different from that of a single GPU. Could you please share your training settings?


Launch args: --multi_gpu --rdzv_backend=c10d
Ext args:    --text_encoder1=/xxx/t5xxl_fp16.safetensors --text_encoder2=/xxx/clip_l.safetensors --flash_attn --split_attn --ddp_timeout=120 --ddp_gradient_as_bucket_view --ddp_static_graph --timestep_sampling=qinglong_flux --guidance_scale=1 --num_timestep_buckets=5 --max_train_epochs=108 --gradient_accumulation_steps=1 --network_args loraplus_lr_ratio=4 --network_dim=32 --network_alpha=16 --network_module=networks.lora_flux --gradient_accumulation_steps=1 --lr_scheduler=cosine_with_min_lr --lr_scheduler_num_cycles=1 --lr_decay_steps=0.2 --lr_scheduler_min_lr_ratio=0.1 --mixed_precision=bf16 --fp8_base --fp8_scaled --persistent_data_loader_workers --img_in_txt_in_offloading --optimizer_type=adamw8bit --save_every_n_epochs=2 --training_comment=flux kontext LoRA model

**Comment by kohya-ss** [comment]
> so readme is Outdated？ can you give guide for multi-GPU training?

I'm working on the documentation :)

**Comment by kohya-ss** [comment]
If you are using a Linux environment, I think it is recommended that you remove `--rdzv_backend=c10d`. Other than that the options seem fine.

In accelerate config, make sure that DDP training is selected and the correct GPU ID is selected. 

It would be great if you could share your config.yaml, which can be found in `~/.cache/huggingface/accelerate/default_config.yaml`.

**Comment by CaiNiaoCYY** [comment]
this is default_config:
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: MULTI_GPU
downcast_bf16: 'no'
enable_cpu_affinity: false
gpu_ids: all
machine_rank: 0
main_process_ip: 127.0.0.1
main_process_port: 8080
main_training_function: main
mixed_precision: bf16
num_machines: 2
num_processes: 8
rdzv_backend: c10d
same_network: false
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false

and machine is 2gpu(A100)

<img width="1680" height="590" alt="Image" src="https://github.com/user-attachments/assets/c93be0bf-f86d-42a3-90ab-9c8bd371919e" />

**Comment by kohya-ss** [comment]
Did you configure it using accelerate config or did you edit the yaml file directly? In my understanding, `num_machines: 2` means multi-node training.

If you're training on one PC with two GPUs, your yaml file should look something like this:
```
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: MULTI_GPU
downcast_bf16: 'no'
enable_cpu_affinity: false
gpu_ids: 0,1
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

**Comment by CaiNiaoCYY** [comment]
1. run accelerate env

Copy-and-paste the text below in your GitHub issue

- `Accelerate` version: 1.6.0
- Platform: Linux-5.4.241-1-tlinux4-0017.4-x86_64-with-glibc2.28
- `accelerate` bash location: /data/miniconda3/envs/env-3.10.6_train_kontext/bin/accelerate
- Python version: 3.10.6
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.4.0+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- PyTorch SDAA available: False
- PyTorch MUSA available: False
- System RAM: 1006.97 GB
- GPU type: NVIDIA A100-SXM4-40GB
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: bf16
        - use_cpu: False
        - debug: False
        - num_processes: 8
        - machine_rank: 0
        - num_machines: 2
        - gpu_ids: all
        - main_process_ip: 127.0.0.1
        - main_process_port: 8080
        - rdzv_backend: c10d
        - same_network: False
        - main_training_function: main
        - enable_cpu_affinity: False
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []

2. image count: 54   repeat: 4   epoch : 150
3. my cmd: python -m accelerate.commands.launch --num_cpu_threads_per_process=8 --multi_gpu  flux_kontext_train.py
4. steps: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8100/8100 [24:39:52<00:00, 10.96s/it, avr_loss=0.325]

2 A100,  train time is 24h, Is this as expected？

**Comment by kohya-ss** [comment]
> 2 A100, train time is 24h, Is this as expected？

The time per step depends on the resolution of the training images. Also, 150 epochs of training seems a long. If the trained LoRA was not overfit after 150 epochs, try increasing the learning rate and reducing the number of epochs.

**Comment by CaiNiaoCYY** [comment]
Thank you for your answer. I will continue to study and take a look.
Also looking forward to your document，Thank you

**Comment by Oruli** [comment]
@kohya-ss @CaiNiaoCYY Unless I'm missing the point of this, I don't believe it's working as expected. There is absolutely no benefit to second/iteration or VRAM usage. In fact it uses the same VRAM across both GPUs as it does on a single GPU with no performance gain?

This has been working on diffusion-pipe for some time, I would be very happy to help test any changes you make to get this working and fast!

My accelerate config:

```
- `Accelerate` version: 1.6.0
- Platform: Linux-6.14.0-29-generic-x86_64-with-glibc2.41
- `accelerate` bash location: /musubi/bin/accelerate
- Python version: 3.12.11
- Numpy version: 2.2.6
- PyTorch version (GPU?): 2.8.0+cu129 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- PyTorch SDAA available: False
- PyTorch MUSA available: False
- System RAM: 123.04 GB
- GPU type: NVIDIA GeForce RTX 5090
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: MULTI_GPU
	- mixed_precision: bf16
	- use_cpu: False
	- debug: False
	- num_processes: 2
	- machine_rank: 0
	- num_machines: 1
	- gpu_ids: all
	- rdzv_backend: static
	- same_network: True
	- main_training_function: main
	- enable_cpu_affinity: False
	- downcast_bf16: no
	- tpu_use_cluster: False
	- tpu_use_sudo: False
	- tpu_env: []
```

**Comment by kohya-ss** [comment]
> [@kohya-ss](https://github.com/kohya-ss) [@CaiNiaoCYY](https://github.com/CaiNiaoCYY) Unless I'm missing the point of this, I don't believe it's working as expected. There is absolutely no benefit to second/iteration or VRAM usage. In fact it uses the same VRAM across both GPUs as it does on a single GPU with no performance gain?

If you use two GPUs for multi-GPU training, the time per step will be the same, but the number of steps should be halved. Please verify this.

**Comment by Oruli** [comment]
> > [@kohya-ss](https://github.com/kohya-ss) [@CaiNiaoCYY](https://github.com/CaiNiaoCYY) Unless I'm missing the point of this, I don't believe it's working as expected. There is absolutely no benefit to second/iteration or VRAM usage. In fact it uses the same VRAM across both GPUs as it does on a single GPU with no performance gain?
> 
> If you use two GPUs for multi-GPU training, the time per step will be the same, but the number of steps should be halved. Please verify this.

Your default console log doesn't even output time per step, it shows s/it?

My comparison from diffusion-pipe is 1GPU = 0.38 samples/s and 2GPU = 0.52 samples/s

Roughly the seconds/iter there is 15, with Musubi it's 25.

If I disable MultiGPU here, steps change from 960 to 1920 indeed, a little confusing though to be honest and I'm still not sure how I can find out what software is faster! I'm assuming if diffusionp-pipe is doing iteration quicker then it would complete faster?

**Comment by kohya-ss** [comment]
> Your default console log doesn't even output time per step, it shows s/it?

When one step is less than one second, 'steps/sec' is displayed. When one step is more than one second, 'sec/step' is displayed.

If Musubi Tuner takes 25 seconds per step when training with 2 GPUs, then it processes two items (two pieces of data) per step, which means it takes 12.5 seconds per item. So it's probably faster than diffusion-pipe's 15 seconds per item.

However, please note that there are other factors that make it difficult to make a general comparison. And also the speed will probably vary depending on other settings.

**Comment by Oruli** [comment]
> > Your default console log doesn't even output time per step, it shows s/it?
> 
> When one step is less than one second, 'steps/sec' is displayed. When one step is more than one second, 'sec/step' is displayed.
> 
> If Musubi Tuner takes 25 seconds per step when training with 2 GPUs, then it processes two items (two pieces of data) per step, which means it takes 12.5 seconds per item. So it's probably faster than diffusion-pipe's 15 seconds per item.
> 
> However, please note that there are other factors that make it difficult to make a general comparison. And also the speed will probably vary depending on other settings.

To be honest that's super confusing! I understand it's a product of not having multigpu training previously.

Would you mind updating the logging logic to factor in multigpu training when you get time? More people use it than you know so it would be super handy for it to be directly comparable to single gpu without having to do calculations and figure out (like here) what's actually going into the calculation, or not.

I would love to see it at 12.5s, but others I'm speaking with have reverted back to diffusion-pipe because they think it's a lot faster
